### ISSUE TITLE
[PodSecurityPolicy] API changes

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#56174]
This is a cover bug to aggregate non-backwards compatible API changes for consideration to the PodSecurityPolicy (i.e. for `v1beta2`).

- [ ] The empty policy should be equivalent to a fully privileged policy. This is a required property of new fields for backwards compatibility, and making all fields behave this way will make the API more consistent. The simplest approach is to default fields to the most permissive value (e.g. `privileged=true`, `allowedCapabilities=*`, etc.). This requires that most fields change to pointer values.

- [ ] All fields should be optional (with the exception of fields with dependencies on other fields, e.g. user MustRunAs), with defaults as described above. The motivation is the same as above, and reduces boilerplate (e.g. for unused features).

- [ ] Organize volume controls. E.g. `AllowedHostPaths`, and [`AllowedFlexVolumes`](https://github.com/kubernetes/kubernetes/pull/53179). As more controls are added for individual volume types, it would be good to have some more structure to these options.

Open for discussion.

**Comment by php-coder** [comment]
CC @adelton @pweil- @liggitt

**Comment by liggitt** [comment]
@kubernetes/sig-auth-proposals

**Comment by mikedanese** [comment]
Security usability issue: RunAsUser and RunAsGroup are disingenuous as alone they allow pod processes to escalate to any user or group (including root) using SUID/SGID. I don't think this is obvious behavior for users of PodSecurityPolicy. Can we modify defaults to disallow this unless explicitly requested?

**Comment by php-coder** [comment]
I found that some types were named inconsistently:
```console
$ grep '^type [^ ]\+ string' pkg/apis/extensions/types.go
type DeploymentStrategyType string
type DeploymentConditionType string
type DaemonSetUpdateStrategyType string
type DaemonSetConditionType string
type ReplicaSetConditionType string
type FSType string
type SELinuxStrategy string
type RunAsUserStrategy string
type FSGroupStrategyType string
type SupplementalGroupsStrategyType string
```
Why `SELinuxStrategy` and `RunAsUserStrategy` don't have `Type` suffix? If it's OK, then why `FSGroupStrategyType` and `SupplementalGroupsStrategyType` have this suffix?

**Comment by pweil-** [comment]
> Why SELinuxStrategy and RunAsUserStrategy don't have Type suffix? If it's OK, then why FSGroupStrategyType and SupplementalGroupsStrategyType have this suffix?

Less a question of why (let's blame me for that one) and more a question of which one to standardize on.  I'd vote for the non-type suffix as it just seems redundant.

**Comment by php-coder** [comment]
> Less a question of why (let's blame me for that one) and more a question of which one to standardize on.

Yes, indeed, it's not so interesting of why because it's already there. My question was -- how it should be named and how to proceed to that? I posed it here, as removal of the old constants looks like a breaking change. On another hand, it should be possible to introduce a new ones types without breaking anything.

> I'd vote for the non-type suffix as it just seems redundant.

First I also thought that the suffixes should be removed, but when I did a grep and saw that Deployments/ReplicaSet have such suffixes I started to doubt.

**Comment by php-coder** [comment]
Another proposed breaking change is to change the default behavior for sysctls -- right now `nil` means "allow all":
https://github.com/kubernetes/kubernetes/blob/a0844c17bfb964e3d1709784203e8404f790116c/pkg/security/podsecuritypolicy/sysctl/mustmatchpatterns.go#L39-L48
I'd suggest to make it stricter by default.

Perhaps, we even can do that change without waiting for major release because sysctl support in Alpha state. WDYT?

CC @sttts

**Comment by tallclair** [comment]
> Another proposed breaking change is to change the default behavior for sysctls

Hmm, so I stated in the description that I think all fields should default to "allow all". I could maybe be convinced that the default could only allow up to the pod default (e.g. capabilities default to allowing the default set of capabilities, as opposed to allowing all capabilities). I'm worried that this would cause problems if defaults change across API versions of Pods though.

**Comment by sttts** [comment]
We only allow sysctls that are "safe", which mainly means that pod isolation is good enough to avoid harm using them. But I don't have a strong opinion about the default in PSPs either.

**Comment by liggitt** [comment]
> The empty policy should be equivalent to a fully privileged policy. This is a required property of new fields for backwards compatibility

That only works if all new fields added to the Pod API are restrictive, which will likely not be the case (c.f. https://github.com/kubernetes/community/pull/1934)

> I stated in the description that I think all fields should default to "allow all". I could maybe be convinced that the default could only allow up to the pod default

That is closer to being workable (assuming new fields in the Pod API default to their current behavior), but is harder to explain

**Comment by liggitt** [comment]
> That is closer to being workable (assuming new fields in the Pod API default to their current behavior), but is harder to explain

Come to think of it, that's more or less what we're doing already:

a new field added to Pods that expands powers:
* defaults to whatever value maintains the current behavior / level of power
* the corresponding control field in PSP defaults to preventing the power expansion (otherwise existing PSPs suddenly allow more powerful pods)

a new field added to Pods that allows limiting powers:
* defaults to whatever value maintains the current behavior / level of power
* the corresponding control field in PSP defaults to not requiring/setting the limitation (otherwise existing PSPs suddently become more restrictive)
